modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/ArtifactLifecycleListenerTestCase.java

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+
+package org.mule.runtime.module.deployment.internal;
+
+import static org.mule.runtime.module.deployment.internal.DeploymentDirectoryWatcher.CHANGE_CHECK_INTERVAL_PROPERTY;
+import static org.mule.runtime.module.deployment.internal.TestArtifactsCatalog.echoTestClassFile;
+import static org.mule.runtime.module.deployment.internal.TestArtifactsCatalog.withLifecycleListenerPlugin;
+import static org.mule.tck.junit4.rule.LogCleanup.clearAllLogs;
+import static org.mule.test.allure.AllureConstants.JavaSdk.ArtifactLifecycleListener.ARTIFACT_LIFECYCLE_LISTENER;
+import static org.mule.test.allure.AllureConstants.JavaSdk.JAVA_SDK;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertThat;
+
+import org.mule.runtime.deployment.model.api.application.Application;
+import org.mule.runtime.deployment.model.api.domain.Domain;
+import org.mule.runtime.module.artifact.api.classloader.ArtifactClassLoader;
+import org.mule.runtime.module.deployment.api.DeploymentListener;
+import org.mule.runtime.module.deployment.impl.internal.builder.ApplicationFileBuilder;
+import org.mule.runtime.module.deployment.impl.internal.builder.DeployableFileBuilder;
+import org.mule.runtime.module.deployment.impl.internal.builder.DomainFileBuilder;
+import org.mule.sdk.api.artifact.lifecycle.ArtifactLifecycleListener;
+import org.mule.tck.junit4.rule.SystemProperty;
+import org.mule.tck.probe.JUnitLambdaProbe;
+import org.mule.tck.probe.PollingProber;
+
+import java.lang.ref.PhantomReference;
+import java.lang.ref.ReferenceQueue;
+
+import io.qameta.allure.Feature;
+import io.qameta.allure.Story;
+import org.junit.Rule;
+import org.junit.Test;
+
+/**
+ * Tests that Extension's {@link ArtifactLifecycleListener}s are honoured for the different lifecycle phases.
+ * <p>
+ * Currently, only artifact disposal is supported.
+ */
+@Feature(JAVA_SDK)
+@Story(ARTIFACT_LIFECYCLE_LISTENER)
+public class ArtifactLifecycleListenerTestCase extends AbstractDeploymentTestCase {
```

Comment:
This test is good as an integration test and we should keep it, but it's too lax when it comes to actually testing the releasers functionality. The SDK feature could break and this test still be green. That is because it's asserting an end state, but not how we got there. ActiveMQ or AWS could fix the underlying bugs making this test go green even if the functionality broke. So this test needs to be complemented with SDK specific functional tests that assert that the releasers are picked up and executed.

Reply:
I'm not sure I totally understood what you mean with the last part. This test uses a synthetic Extension which purposely leaks some ClassLoaders by spawning some Threads and leaving them running forever (until the `ArtifactLifecycleListener` of the Extension stops them). I don't see how the ActiveMQ or AWS releasers could make this test go green.

That being said, I agree that I should add more tests covering the correctness of all the `ArtifactDisposalContext` methods. This test falls a little bit short on that.

Reply:
The unit test for the `DefaultArtifcatDisposalContext` implementation has been added, [here](https://github.com/mulesoft/mule/pull/12421/files#diff-ee94482878114a432879c8b64d847ca6b32267106c61279c866fe08cdda2862c).